src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -494,6 +494,35 @@ def upgrade_stop(self) -> str:
         self.mgr.event.set()
         return 'Stopped upgrade to %s' % target_image
 
+    def upgrade_service(self, service_type: str, service_image: str, image: str) -> List[str]:
+        key = f'mgr/cephadm/container_image_{service_image}'
+        try:
+            self.mgr.check_mon_command({
+                'prefix': 'config set',
+                'name': key,
+                'value': image,
+                'who': 'mgr',
+            })
+        except MonCommandFailed as e:
+            self.mgr.log.exception(f"Failed to set {service_image} option {key}")
+            raise OrchestratorError(f"Failed to set {service_image} option {key}: {e}")
+
+        out = []
+        # Redeploy all the service daemons of provided service type
+        for service_name, spec in self.mgr.spec_store.all_specs.items():
+            if service_type != spec.service_type:
+                continue
```

Comment:
rather than this looping over all the specs, we should probably add some sort of `get_specs_by_type` function to the SpecStore class similar to how we have a `get_daemons_by_type` for the HostCache class.